### PR TITLE
[Frontend] add `add_request_id` middleware

### LINKED ISSUE #9593: [Feature]: support `x-request-id` header
### 🚀 The feature, motivation and pitch

Related: #9550 

It is a common approach for server admin to backtrack client requests via a "request id"(cf. [this SE post](https://stackoverflow.com/questions/25433258/what-is-the-x-request-id-http-header)). It is a unique identifier assigned to each HTTP request which helps grepping the request info from server log.

OpenAI supports this in the form of response header (cf. [API Reference - OpenAI API](https://platform.openai.com/docs/api-reference/debugging-requests)). Any response from OpenAI API server has `x-request-id` header, and `openai-python` package provides a convenient way to retrieve this header. (cf. [openai/_models.py](https://github.com/openai/openai-python/blob/main/src/openai/_models.py#L105-L120))

In the referenced PR, we observed the need for request identification (a typical usecase of `x-request-id`), so it may be a good time to support this optional HTTP header in online serving.

Suggestion:

- Each API response is sent with a `x-request-id` header
- When `x-request-id` header is given in user request, send it back to the response header ("Idempotency")
- Otherwise, `x-request-id` is a random uuid hex value ("compatibility with OpenAI API behavior")

Demo:

Case 1. `X-Request-Id` header is not given - return random hex
```console
$ curl -v -X POST http://localhost:8000/v1/chat/completions \
> -d '{"model":"meta-llama/Llama-3.2-1B-Instruct","messages":[{"role":"user","content":"Hi"}]}' \
> -H 'Content-Type: Application/json'
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> POST /v1/chat/completions HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.68.0
> Accept: */*
> X-Request-Id: aaaa
> Content-Type: application/json
> Content-Length: 88
> 
< HTTP/1.1 200 OK
< date: Tue, 22 Oct 2024 15:58:16 GMT
< server: uvicorn
< content-length: 230
< content-type: application/json
< x-request-id: 26970bda2e124bb8ad293d3d25d8a4cf
...
```

Case 2. `X-Request-Id` header is specified - pass it back
```console
$ curl -v -X POST http://localhost:8000/v1/chat/completions \
> -d '{"model":"meta-llama/Llama-3.2-1B-Instruct","messages":[{"role":"user","content":"Hi"}]}' \
> -H 'X-Request-Id: aaaa' \
> -H 'Content-Type: Application/json'
*   Trying 127.0.0.1:8000...
* Connected to localhost (127.0.0.1) port 8000 (#0)
> POST /v1/chat/completions HTTP/1.1
> Host: localhost:8000
> User-Agent: curl/7.68.0
> Accept: */*
> X-Request-Id: aaaa
> Content-Type: application/json
> Content-Length: 88
> 
< HTTP/1.1 200 OK
< date: Tue, 22 Oct 2024 15:58:16 GMT
< server: uvicorn
< content-length: 230
< content-type: application/json
< x-request-id: aaaa
...
```

### Alternatives

#9550 achieves similar goal, but this PR has some pros:

- this PR lies within OpenAI API spec while #9550 extends the API spec (possibility in API conflict in the future?)
- OpenAI SDKs supports `x-request-id` in the way that returns the exact value set by a client (e,g, `response._request_id` attribute in `openai-python`), so no post-processing needed
- HTTP header is easier to handle than HTTP body, e.g. it is easier to grep request id using `curl` cmd etc.

### Additional context

Some other language model APIs send request id ([anthropic](https://docs.anthropic.com/en/api/errors#request-id), [hyperclova](https://api.ncloud-docs.com/docs/en/clovastudio-chatcompletions))

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by cjackal**
FIX #9593 

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by cjackal** [comment]
It's a little vague where and how unit test should be written, suggestions are welcome.

File: vllm/entrypoints/openai/api_server.py [hunk 1]
```diff
@@ -449,6 +450,13 @@ async def authentication(request: Request, call_next):
                                     status_code=401)
             return await call_next(request)
 
+    @app.middleware("http")
+    async def add_request_id(request: Request, call_next):
+        request_id = request.headers.get("X-Request-Id", uuid.uuid4().hex)
+        response = await call_next(request)
+        response.headers["X-Request-Id"] = request_id
+        return response
```
Comment (guoyuhong) [inline vllm/entrypoints/openai/api_server.py:458]: This `X-Request-id` seems not connected to `request_id` in (`create_chat_completion`)[https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_chat.py#L179].

Comment (cjackal) [reply vllm/entrypoints/openai/api_server.py:458]: Right, this id has nothing to do with chat request id(or request id of other OpenAI API response body). This behavior is consistent with OprnAI's current behavior, and you can recover this id like:

```python
from openai import OpenAI

client = OpenAI(...)
response = client.chat.completions.create(...)
response._request_id
```

Comment (guoyuhong) [reply vllm/entrypoints/openai/api_server.py:458]: Then, LGTM.

File: vllm/entrypoints/openai/api_server.py [hunk 2]
```diff
@@ -449,6 +450,13 @@ async def authentication(request: Request, call_next):
                                     status_code=401)
             return await call_next(request)
 
+    @app.middleware("http")
+    async def add_request_id(request: Request, call_next):
+        request_id = request.headers.get("X-Request-Id", uuid.uuid4().hex)
```
Comment (simon-mo) [inline vllm/entrypoints/openai/api_server.py:455]:
```suggestion
        request_id = request.headers.get("X-Request-Id") or uuid.uuid4().hex
```
small optimization, The previous version will evaluate uuid4() before getting the item from headers

**Review by simon-mo — APPROVED** [review]
This makes sense to me given it's industry standard. It would be great to add comment about this as well. Thanks!

**Comment by cjackal** [comment]
@simon-mo Just have added an example using custom headers in the `openai-compatible-server` page (and sign-off), tell me if there's better spot to add comment about it.

File: docs/source/serving/openai_compatible_server.md [hunk 3]
```diff
@@ -35,14 +35,17 @@ vLLM also provides experimental support for OpenAI Vision API compatible inferen
 ## Extra Parameters
 vLLM supports a set of parameters that are not part of the OpenAI API.
 In order to use them, you can pass them as extra parameters in the OpenAI client.
-Or directly merge them into the JSON payload if you are using HTTP call directly.
+Or directly merge them into the HTTP headers or JSON payload if you are using HTTP call directly.
```
Comment (DarkLight1337) [inline docs/source/serving/openai_compatible_server.md:38]: It would be better to mention that only request ID can be set via HTTP headers.

Comment (DarkLight1337) [reply docs/source/serving/openai_compatible_server.md:51]: Please also revert this paragraph.

Comment (cjackal) [reply docs/source/serving/openai_compatible_server.md:51]: Great, I'll add a new section for extra HTTP headers.

File: docs/source/serving/openai_compatible_server.md [hunk 4]
```diff
@@ -62,6 +62,32 @@ completion = client.chat.completions.create(
...
+Only `X-Request-Id` HTTP request header is supported for now.
+
+```python
+completion = client.chat.completions.create(
+  model="NousResearch/Meta-Llama-3-8B-Instruct",
+  messages=[
+    {"role": "user", "content": "Classify this sentiment: vLLM is wonderful!"}
+  ],
+  extra_headers={
+    "x-request-id": "sentiment-classification-00001",
```
Comment (DarkLight1337) [inline docs/source/serving/openai_compatible_server.md:76]: Just to be sure: does the capitalization of `X-Request-Id` matter?

Comment (cjackal) [reply docs/source/serving/openai_compatible_server.md:76]: It doesn't; `openai-python` library uses `httpx` for its HTTP client SDK, and `httpx` handles the capitalization part.

**Review by DarkLight1337 — APPROVED** [review]
Looks good then, sorry for the delay!

**Comment by tomeras91** [comment]
This PR made a starlette bug manifest itself in all use-cases of deploying the vLLM OpenAI compatible API server. The bug causes requests cancelled by the client to not be aborted from the LLM engine. Discussed [here](https://github.com/vllm-project/vllm/issues/10087)

**Pull request merged**